faculty/clients/experiment.py

```diff
@@ -245,6 +334,98 @@ def make_restore_runs_response(self, data):
         return RestoreExperimentRunsResponse(**data)
 
 
+class SingleFilterValueField(fields.Field):
+    """
+    Field that serialises/deserialises a run filter.
+    """
+
+    def _is_valid_uuid(self, value, obj):
+        return isinstance(value, uuid.UUID) and (
+            obj.by in {SingleFilterBy.PROJECT_ID, SingleFilterBy.RUN_ID}
+        )
+
+    def _is_valid_experiment_id(self, value, obj):
+        return (
+            isinstance(value, int) and obj.by == SingleFilterBy.EXPERIMENT_ID
+        )
+
+    def _is_directly_stringifiable(self, value, obj):
+        return (
+            self._is_valid_uuid(value, obj)
+            or self._is_valid_experiment_id(value, obj)
+            or obj.by
+            in {
+                SingleFilterBy.TAG,
+                SingleFilterBy.PARAM,
+                SingleFilterBy.METRIC,
+            }
+        )
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        return value
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        if self._is_directly_stringifiable(value, obj):
+            return str(value)
+        elif obj.by == SingleFilterBy.DELETED_AT:
+            return marshmallow_utils.from_iso_datetime(str(value)).isoformat()
+        else:
+            raise RunQueryFilterValidation(
```

Comment:
Why not just raise a marshmallow `ValidationError`? I feel this would be more idiomatic.

Reply:
In general, marshmallow does not validate on serialization, only on deserialization, so it's not obvious tha a `ValidationError` is the correct thing here as it is (by use, if not by definiton) only applicable ot deserialization.

Reply:
To be honest I don't really know what the right thing to do is here. Elsewhere in my additions I have just used a `ValueError` (e.g. in the constructors of Filters). Here I wanted to highlight that it's a serialisation error, but I am happy to change it if anyone has strong thoughts